xbmc/cores/VideoPlayer/DVDDemuxers/DemuxTimeline.cpp

```diff
@@ -0,0 +1,297 @@
+{
+  --chapter;
+  if (chapter < 0 || unsigned(chapter) >= m_chapters.size())
+    return false;
+  CLog::Log(LOGDEBUG, "TimelineDemuxer: Switch Demuxer");
+  m_curChapter = &m_chapters[chapter];
+  bool result = m_curChapter->demuxer->SeekTime(m_curChapter->startSrcTime, true, startpts);
+  if (result && startpts)
+    (*startpts) += m_curChapter->shiftTime();
+  return result;
+}
+
+int CDemuxTimeline::GetChapterCount()
+{
+  return m_chapters.size();
+}
+
+int CDemuxTimeline::GetChapter()
+{
+  return m_curChapter->index + 1;
+}
+
+void CDemuxTimeline::GetChapterName(std::string& strChapterName, int chapterIdx)
+{
+  --chapterIdx;
+  if (chapterIdx < 0 || unsigned(chapterIdx) >= m_chapters.size())
+    return;
+  strChapterName = m_chapters[chapterIdx].title;
+}
+
+int64_t CDemuxTimeline::GetChapterPos(int chapterIdx)
+{
+  --chapterIdx;
+  if (chapterIdx < 0 || unsigned(chapterIdx) >= m_chapters.size())
+    return 0;
+  return (m_chapters[chapterIdx].startDispTime + 999) / 1000;
+}
+
+void CDemuxTimeline::SetSpeed(int iSpeed)
+{
+  for (auto &demuxer : m_demuxer)
+    demuxer->SetSpeed(iSpeed);
+}
+
+int CDemuxTimeline::GetStreamLength()
+{
+  return m_chapterMap.rbegin()->first;
+}
+
+std::vector<CDemuxStream*> CDemuxTimeline::GetStreams() const
+{
+  return m_primaryDemuxer->GetStreams();
+}
+
+int CDemuxTimeline::GetNrOfStreams() const
+{
+  return m_primaryDemuxer->GetNrOfStreams();
+}
+
+std::string CDemuxTimeline::GetFileName()
+{
+  return m_primaryDemuxer->GetFileName();
+}
+
+void CDemuxTimeline::EnableStream(int id, bool enable)
+{
+  for (auto &demuxer : m_demuxer)
+    demuxer->EnableStream(demuxer->GetDemuxerId(), id, enable);
+}
+
+CDemuxStream* CDemuxTimeline::GetStream(int iStreamId) const
+{
+  return m_primaryDemuxer->GetStream(m_primaryDemuxer->GetDemuxerId(), iStreamId);
+}
+
+std::string CDemuxTimeline::GetStreamCodecName(int iStreamId)
+{
+  return m_primaryDemuxer->GetStreamCodecName(m_primaryDemuxer->GetDemuxerId(), iStreamId);
+}
+
+
+CDemuxTimeline* CDemuxTimeline::CreateTimeline(CDVDDemux *demuxer)
+{
+  return CreateTimelineFromMatroskaParser(demuxer);
+  return CreateTimelineFromEbml(demuxer);
+}
+
+std::string segUidToHex(std::string uid)
+{
+	const char *hex = "0123456789abcdef";
+	std::string result("0x");
+	result.reserve(18);
+	for (unsigned char twoDigits : uid)
+	{
+		result.append(1, hex[(twoDigits >> 4) & 0xf]);
+		result.append(1, hex[twoDigits & 0xf]);
+	}
+	return result;
+}
+
+CDemuxTimeline* CDemuxTimeline::CreateTimelineFromMatroskaParser(CDVDDemux *primaryDemuxer)
+{
+  std::unique_ptr<CDVDInputStreamFile> inStream(new CDVDInputStreamFile(CFileItem(primaryDemuxer->GetFileName(), false)));
+  if (!inStream->Open())
+    return nullptr;
+  CDVDInputStream *input = inStream.get();
+
+  MatroskaFile mkv;
+  bool result = mkv.Parse(input);
+  if (!result)
+    return nullptr;
+
+   // multiple editions unsupported (at the moment)
+  if (mkv.segment.chapters.editions.size() != 1)
+    return nullptr;
```

Comment:
I need to be sure i have at least one edition.
Maybe i have parsed a mkv that doesn't define chapters at all. In that case, the following code would cause a segfault.